File: Practice/lehman/Lehman.java
public class Lehman {
    int iterate;
    ModPow modPow;

    public Lehman(int iterate) {
        this.iterate = iterate;
        modPow = new ModPow();
    }

    public boolean isPrime(int p)
    {
        if(p==2 || p==3) return true;

        for(int i=0; i<iterate; i++)
        {
            int a = (int)(Math.random()*(p-3)) +2;

            int ans = modPow.modPow(a, (p-1)/2, p);
            if(ans != 1 && ans != p-1)
            {
                System.out.println("Iteration: "+i);
                return false;
            }
        }
        return true;
    }
}

File: Practice/lehman/lehman.cpp
#include <bits/stdc++.h>
#define ll long long
#define ull unsigned long long
using namespace std;

int iteration = 20;

int modpow(int a, int n, int p)
{
    if (n == 1)
        return a;

    int res = modpow(a, n / 2, p);
    res = (res * res) % p;

    if (n % 2 == 1)
        res = (res * a) % p;
    return res;
}

string lehman(int p)
{
    for (int i = 0; i < iteration; i++)
    {
        int a = ((rand() / RAND_MAX) * (p - 3)) + 2;

        int ans = modpow(a, (p - 1) / 2, p);
        if (ans != 1 && ans != p - 1)
        {
            return "NO";
        }
    }
    return "YES";
}

int main()
{
    int p;
    cin >> p;

    cout << p << " is prime? " << lehman(p) << endl;
}